devito/symbolics/manipulation.py

```diff
@@ -172,21 +199,34 @@ class UxreplaceRegistry(list):
     one such Reconstructable object is encountered.
     """
 
-    def register(self, cls, rkwargs_callback_mapper=None):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.deep_cls = []  # Classes only touched by deep uxreplace
+
+    def register(self, cls, rkwargs_callback_mapper=None, deep=False):
+        if deep:
+            self.deep_cls.append(cls)
         self.append(cls)
         _uxreplace_handle.register(cls, _uxreplace_handle_reconstructable)
 
         for kls, callback in (rkwargs_callback_mapper or {}).items():
             _uxreplace_dispatch.register(kls, callback)
 
-    def dispatchable(self, obj):
-        return isinstance(obj, tuple(self))
+    def dispatchable(self, obj, deep=False):
```

Comment:
So, I'm not sure I fully understand...

For `ConditionalDimension`:
* sometimes you want it `deep` 
* but some other times, you don't ?

is this correct? in particular, you want it deep **only** when you calol it inside `separate_dimensions`?

Reply:
Inside `separate_dimensions`, `uxreplace` wants to recurse into `ConditionalDimension`s and update all their args, kwargs, etc as necessary. However, elsewhere in the code, in particular [here](https://github.com/devitocodes/devito/blob/7c155e0f852afc15a493d1eacbc05cd706276b85/devito/ir/equations/algorithms.py#L143) when called from [here](https://github.com/devitocodes/devito/blob/7c155e0f852afc15a493d1eacbc05cd706276b85/devito/ir/equations/equation.py#L193), this behaviour may not be desired. In the above case, the `ConditionalDimension` may have already been indexified and shifted during the lowering of the expressions if it contains terms present in the `Eq` (in the case that `ConditionalDimension` is always affected by `uxreplace`), causing shifting to get applied twice. Hence, the need to be able to have some kind of control over the classes touched by `uxreplace` on a call-by-call basis